Python/traceback.c

```diff
@@ -708,15 +744,18 @@ tb_displayline(PyTracebackObject* tb, PyObject *f, PyObject *filename, int linen
                                 filename, lineno, name);
     if (line == NULL)
         return -1;
-    err = PyFile_WriteObject(line, f, Py_PRINT_RAW);
+    err = _Py_WriteIndentedMargin(margin_indent, margin, f);
+    err |= PyFile_WriteObject(line, f, Py_PRINT_RAW);
     Py_DECREF(line);
     if (err != 0)
         return err;
```

Comment:
> > It's a mess, so how about I make a PR to clean up the current code's error checking, we merge that, then I recreate this one.

Change of plans - I found that its more delicate than I originally thought (see bpo-45635). I don't want to hold up this PR until that's sorted, so I'll make sure I'm not adding more error checking issues in the code I'm adding here, but leave the rest for later.

Reply:
I added two limits  max_group_depth and max_group_width.   If we roll everything into one limit I think we will get a constant stream of bug reports about edge cases where you got the wrong part of the output.  Though it would probably be rare that it makes a difference in practice. 

I didn't add a limit on the length of context/cause chain, because that's outside the scope of this PR. I created bpo-45694 for that.